### PR TITLE
Closes #4438: Add support for browser.tabs.create()

### LINKED ISSUE #4438: Expose WebExtensionController.TabDelegate
GeckoView added [WebExtensionController.TabDelegate](https://mozilla.github.io/geckoview/javadoc/mozilla-central/org/mozilla/geckoview/WebExtensionController.TabDelegate.html) that adds support for web extensions APIs like `browser.tabs.remove` and `browser.tabs.create`

### PR DESCRIPTION
**Pull request by psymoon**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtensionsTabsDelegate.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.concept.engine.webextension
+
+import mozilla.components.concept.engine.EngineSession
+
+/**
+ * Provides delegation for when a extension opens or closes a tab.
```
Comment (csadilek) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtensionsTabsDelegate.kt]: This is a bit mechanic to read, plus we're not supporting "closing a tab". Let's file a follow-up for the latter and add a link.

I'd change the KDocs to something like: "Notifies applications / other components that a web extension wants to open a new tab via browser.tabs.create. Note that browser.tabs.remove is currently not supported: link to issue...

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtensionsTabsDelegate.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ */
+interface WebExtensionsTabsDelegate {
+
+    /**
+     * Invoked when a web extension opens a new tab.
+     *
+     * @param webExtension An instance of [WebExtension] or null if extension was not registered
```
Comment (csadilek) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtensionsTabsDelegate.kt]: nit: was not registered with the engine.

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/Engine.kt [hunk 3]
```diff
@@ -129,6 +130,15 @@ interface Engine {
         onError: ((String, Throwable) -> Unit) = { _, _ -> }
     ): Unit = onError(id, UnsupportedOperationException("Web extension support is not available in this engine"))
 
+    /**
+     * Registers for web extension tab delegation.
```
Comment (csadilek) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/Engine.kt]: nit: Registers a [WebExtensionTabDelegate] to be notified when web extensions attempt to open/close tabs.

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtensionsTabsDelegate.kt [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.concept.engine.webextension
+
+import mozilla.components.concept.engine.EngineSession
+
+/**
+ * Provides delegation for when a extension opens or closes a tab.
+ */
+interface WebExtensionsTabsDelegate {
```
Comment (csadilek) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtensionsTabsDelegate.kt]: nit: The double plural is a bit hard to read. Let's refactor to just `WebExtensionTabDelegate` similar to GeckoView.

File: docs/changelog.md [hunk 5]
```diff
@@ -27,6 +27,18 @@ permalink: /changelog/
 * **feature-intent**
   * Added support for NFC tag intents to `TabIntentProcessor`.
 
+* **engine**, **engine-gecko-nightly** and **engine-gecko-beta**
+  * Added `WebExtensionsTabsDelegate` that supports `browser.tabs.create()`.
```
Comment (csadilek) [inline docs/changelog.md]: nit: maybe "to support `browser.tabs.create()` in web extensions`...

**Comment by psymoon** [comment]
bors r=csadilek

**Comment by Amejia481** [comment]
@psymoon  could we squash the commits? :)

**Comment by psymoon** [comment]
bors r=csadilek

**Comment by Amejia481** [comment]
bors retry

**Comment by psymoon** [comment]
bors r=csadilek

**Comment by psymoon** [comment]
bors r=csadilek

**Pull request merged**